legend-engine-xt-changetoken-compile/src/main/java/org/finos/legend/engine/changetoken/generation/GenerateCastUtil.java

```diff
@@ -0,0 +1,99 @@
+//  Copyright 2022 Goldman Sachs
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package org.finos.legend.engine.changetoken.generation;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+public class GenerateCastUtil
+{
+    public static Object resolveRelativeReference(Map<String, Object> rootObjectNode, List<Object> path, String relativeReference)
+    {
+        List<Object> newPath = resolvePath(path, relativeReference);
+
+        /*
+            extract the value by path
+         */
+        Object cur = rootObjectNode;
+        for (Object key : newPath)
+        {
+            if (cur instanceof Map)
+            {
+                cur = ((Map<String, Object>) cur).get((String) key);
+            }
+            else if (cur instanceof List)
+            {
+                cur = ((List<Object>) cur).get((Integer) key);
+            }
+            else
+            {
+                throw new RuntimeException("Unexpected key on a non-container object");
+            }
+        }
+        return cur;
+    }
+
+    public static void setRelativeReference(Map<String,Object> rootObjectNode, List<Object> path, String relativeReference, Object value)
+    {
+        List<Object> newPath = resolvePath(path, relativeReference);
+
+        /*
+            set the value by path
+         */
+        Object cur = rootObjectNode;
+        for (Object key : newPath.subList(0, newPath.size() - 1))
+        {
+            if (cur instanceof Map)
+            {
+                cur = ((Map<Object, Object>) cur).get((String) key);
+            }
+            else if (cur instanceof List)
+            {
+                cur = ((List<Object>) cur).get((Integer) key);
```

Comment:
Since relativeReference is a String, I think you might need to try parsing key to an Integer.
```suggestion
                cur = ((List<Object>) cur).get((key instanceof Integer) ? (Integer) key : Integer.valueOf(key));
```